docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+  oneof qux {
+    string quux = 3;
+  }
+  repeated string many = 4;
+  map<string, string> mapping = 5;
+}
+```
+
+```proto
+// service.proto
+
+import "custom_options.proto";
+
+service FooService {
+  rpc Get(GetRequest) returns (GetResponse) {
+    option (service_method_option) = { 
+        foo: 567, 
+        bar: "Some string", 
+        quux: "Oneof string",
+        many: ["a", "b", "c"],
+        mapping: [{key: "testKey", value: "testVal"}]
+    };
+  }
+}
+```
+
+You can retrieve the options using a generated type by first generating the file which defines the custom option type.  Then, import and pass this type to the `findCustomMessageOption` function.
+
+```ts
+import { ServiceOptions } from "./gen/proto/custom_options_pb.js";
+
+const option = findCustomMessageOption(method, 50007, ServiceOptions)
+
+console.log(option);
+/*
+ * {
+ *     foo: 567, 
+ *     bar: "Some string", 
+ *     quux: "Oneof string",
+ *     many: ["a", "b", "c"],
+ *     mapping: [{key: "testKey", value: "testVal"}]
+ * }
+ */
+ ```
+ 
+ Alternatively, you can use a type created at runtime:
+
+```ts
+const opts = proto3.makeMessageType("ServiceOptions", [
+  {
+    no: 1,
+    name: "foo",
+    kind: "scalar",
+    T: ScalarType.STRING,
+  },
+  {
+    no: 2,
+    name: "bar",
+    kind: "scalar",
+    T: ScalarType.STRING,
+  },
+  {
+    no: 3,
+    name: "quux",
+    kind: "scalar",
+    T: ScalarType.STRING,
+  },
+  {
+    no: 4,
+    name: "many",
+    kind: "scalar",
+    repeated: true,
+    T: ScalarType.STRING,
+  },
+  {
+    no: 5,
+    name: "mapping",
+    kind: "map",
+    K: ScalarType.STRING,
+    V: {
+      kind: "scalar",
+      T: ScalarType.STRING,
+    },
+  },
+]);
+const option = findCustomMessageOption(method, 50007, opts)
+
+console.log(option);
+/**
+  {
+      foo: 567, 
+      bar: "Some string", 
+      quux: "Oneof string",
+      many: ["a", "b", "c"],
+      mapping: [{key: "testKey", value: "testVal"}]
+  }
+ */
+```
+
+Note that `repeated` and `map` values are only supported within a custom message option.  They are not supported as option types independently.  If you have need to use a custom option that is `repeated` or is of type `map`, it is recommended to use a message option as a wrapper.
+
+
+#### Enum Options
+
+Custom options of an enum type can be retrieved via the `findCustomEnumOption` function.  It returns a `number` corresponding to the enum value set in the option.
+
+```ts
+export function findCustomEnumOption(
+  desc: AnyDesc,
+  id: number
+): number | undefined {
+```
+
+For example, given the following:
+
+```proto
+extend google.protobuf.MessageOptions {
+  optional FooEnum foo_enum_option = 50001;
+}
+
+enum FooEnum {
+  UNDEFINED = 0;
+  ACTIVE = 1;
+  INACTIVE = 2;
+}
+
+message FooMessage {
+  option (foo_enum_option) = ACTIVE;
+
+  string name = 1;
+}
+```
+
+The value of this option can be retrieved as follows:
+
+```ts
+const enumVal = findCustomEnumOption(descMessage, 50001);  // 1
```

Comment:
If we recommend to generate enumerations used in custom options, like we do for messages, this could be:

```ts
const enumVal: ServiceStatus | undefined = findCustomEnumOption(service, 50006);
```

What do you think?